flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/vertica/VerticaDialect.java

```diff
@@ -0,0 +1,109 @@
+
+import java.util.EnumSet;
+import java.util.Optional;
+import java.util.Set;
+
+/** JDBC dialect for Vertica. */
+public class VerticaDialect extends AbstractDialect {
+
+    private static final long serialVersionUID = 1L;
+
+    // Define MAX/MIN precision of TIMESTAMP type according to Vertica docs:
+    // https://www.vertica.com/docs/12.0.x/HTML/Content/Authoring/SQLReferenceManual/DataTypes/Date-Time/DateTimeDataTypes.htm
+    private static final int MAX_TIMESTAMP_PRECISION = 6;
+    private static final int MIN_TIMESTAMP_PRECISION = 0;
+
+    // Define MAX/MIN precision of DECIMAL type according to Vertica docs:
+    // https://www.vertica.com/docs/12.0.x/HTML/Content/Authoring/SQLReferenceManual/DataTypes/Numeric/NUMERIC.htm
+    private static final int MAX_DECIMAL_PRECISION = 1024;
+    private static final int MIN_DECIMAL_PRECISION = 1;
+
+    @Override
+    public Set<LogicalTypeRoot> supportedTypes() {
+        // List of Vertica data types:
+        // https://www.vertica.com/docs/12.0.x/HTML/Content/Authoring/SQLReferenceManual/DataTypes/SQLDataTypes.htm
+        // https://www.vertica.com/docs/12.0.x/HTML/Content/Authoring/ConnectingToVertica/ClientJDBC/JDBCDataTypes.htm
+
+        return EnumSet.of(
+                LogicalTypeRoot.BINARY,
+                LogicalTypeRoot.VARBINARY,
+                LogicalTypeRoot.BOOLEAN,
+                LogicalTypeRoot.CHAR,
+                LogicalTypeRoot.VARCHAR,
+                LogicalTypeRoot.DATE,
+                LogicalTypeRoot.TIME_WITHOUT_TIME_ZONE,
+                LogicalTypeRoot.TIMESTAMP_WITH_TIME_ZONE,
+                LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE,
+                LogicalTypeRoot.DOUBLE,
+                LogicalTypeRoot.DECIMAL,
+                LogicalTypeRoot.BIGINT,
+                LogicalTypeRoot.ARRAY);
+    }
+
+    @Override
+    public String dialectName() {
+        return "Vertica";
+    }
+
+    @Override
+    public JdbcRowConverter getRowConverter(RowType rowType) {
+        return new VerticaRowConverter(rowType);
+    }
+
+    @Override
+    public String getLimitClause(long limit) {
+        return "LIMIT " + limit;
+    }
+
+    @Override
+    public Optional<String> defaultDriverName() {
+        return Optional.of("com.vertica.jdbc.Driver");
+    }
+
+    @Override
+    public String quoteIdentifier(String identifier) {
+        return "\"" + identifier + "\"";
+    }
+
+    @Override
+    public Optional<Range> decimalPrecisionRange() {
+        return Optional.of(Range.of(MIN_DECIMAL_PRECISION, MAX_DECIMAL_PRECISION));
+    }
+
+    @Override
+    public Optional<Range> timestampPrecisionRange() {
+        return Optional.of(Range.of(MIN_TIMESTAMP_PRECISION, MAX_TIMESTAMP_PRECISION));
+    }
+
+    @Override
+    public Optional<String> getUpsertStatement(
+            String tableName, String[] fieldNames, String[] uniqueKeyFields) {
+        throw new UnsupportedOperationException(
```

Comment:
we could use a MERGE no?
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Statements/MERGE.htm

see OracleDialect for example..

Reply:
Even though Vertica does support the MERGE statement, I wasn't able to have it working with Flink. This was the implementation I originally did, based on the OracleDialect example:
```java
@Override
public Optional<String> getUpsertStatement(
        String tableName, String[] fieldNames, String[] uniqueKeyFields) {

    String valuesBinding =
            Arrays.stream(fieldNames)
                    .map(f -> ":" + f + " " + quoteIdentifier(f))
                    .collect(Collectors.joining(", "));

    String usingClause = String.format("SELECT %s", valuesBinding);

    String onClause =
            Arrays.stream(uniqueKeyFields)
                    .map(f -> "t." + quoteIdentifier(f) + "=s." + quoteIdentifier(f))
                    .collect(Collectors.joining(" AND "));

    final Set<String> uniqueKeyFieldsSet =
            Arrays.stream(uniqueKeyFields).collect(Collectors.toSet());
    String updateClause =
            Arrays.stream(fieldNames)
                    .filter(f -> !uniqueKeyFieldsSet.contains(f))
                    .map(f -> quoteIdentifier(f) + " = s." + quoteIdentifier(f))
                    .collect(Collectors.joining(", "));

    String insertFields =
            Arrays.stream(fieldNames)
                    .map(this::quoteIdentifier)
                    .collect(Collectors.joining(", "));

    String valuesClause =
            Arrays.stream(fieldNames)
                    .map(f -> "s." + quoteIdentifier(f))
                    .collect(Collectors.joining(", "));

    String mergeQuery =
            "MERGE INTO %s t USING (%s) s "
                    + "ON %s "
                    + "WHEN MATCHED THEN UPDATE SET %s "
                    + "WHEN NOT MATCHED THEN INSERT (%s) "
                    + "VALUES (%s)";

    return Optional.of(
            String.format(
                    mergeQuery,
                    quoteIdentifier(tableName),
                    usingClause,
                    onClause,
                    updateClause,
                    insertFields,
                    valuesClause));
}
```
The constructed MERGE query was syntactically correct, which I have validated by running it on the Vertica database instance (I had to replace :placeholders with the actual values). 
When running the query within Flink(upsert statement) I got this exception, originating from the Vertica JDBC driver:
`java.sql.SQLException: [Vertica][VJDBC](3376) ERROR: Failed to find conversion function from unknown to int`
After trying to find out what was the reason for the exception, I've stumbled upon this StackOverflow post:
https://stackoverflow.com/questions/18073901/failed-to-find-conversion-function-from-unknown-to-text.
 
My guess is that parametrized MERGE statement is not supported by Vertica JDBC driver and that could also be the reason KafkaConnect JDBC connector doesn't support upsert statement with Vertica as well, but only insert.